metadata/src/main/java/org/apache/kafka/controller/PartitionReassignmentReplicas.java

```diff
@@ -124,6 +124,8 @@ Optional<CompletedReassignment> maybeCompleteReassignment(List<Integer> targetIs
             if (!newTargetIsr.contains(replica)) return Optional.empty();
         }
 
+        if (newTargetIsr.size() < Math.min(minIsrCount, newTargetReplicas.size())) return Optional.empty();
```

Comment:
I haven't been following along the ELR KIP closely so I may be missing some of the details. 
My understanding is that a reassignment is only completed once the target replicas are all in the ISR. Isn't it better to just reject the initiation of a reassignment where the target replica count is less than the configured ISR?

From what I can tell, if a user initiates a reassignment where the target replica count < ISR, the reassignment will just get stuck indefinitely.

Reply:
I think it is more like a design issue/bug. The following example is valid currently:
An existing replica set is [0,1,2], and a new partition assignment is [0,1,3]. So the "adding" replica is 3. The controller can finish the reassignment if the ISR if the current ISR is [3]. 
This PR prevents completing such reassignment if min ISR is 2.

Reply:
A better example is with min ISR=3
1. originally replicas [0,1,2,3], ISR[0,1,2,3]
2. Start reassignment. Target replica set [0,1,7,8]. Temporarily, the replica set is [0,1,2,3,7,8], ISR[0,1,2,3].
3. Replication goes on and some brokers fail and up, and we could experience the following transition for ISR. [0,1,2,3] -> [3,7] -> [7] -> [7,8]
4. At this point, the reassignment can be completed.
5. In the current world, we can't do much for the above. But in the ELR world, we can have something different in step 3.

ISR/ELR: [0,1,2,3]/[] -> [3,7]/[2] -> [7]/[2,3] -> [7,8]/[2,3]. If we complete the reassignment now, we loss the potential good replicas [2,3] and have a higher risk of data loss.

Reply:
@CalvinConfluent 
hmm, interesting. this is an inconsistency in the KRaft implementation of reassignment completion and the ZK controller implementation of reassignment completion.

In the ZK controller, we only allow a reassignment to be completed if the ISR contains all of the target replicas. So in your second example, the ISR would need to contain [0, 1, 7, 8].

code reference is in `KafkaController.isReassignmentComplete`:
```
        val targetReplicas = assignment.targetReplicas.toSet
        targetReplicas.subsetOf(isr)
```

I am wondering whether it was intentional to change the criteria? IMO, the ZK controller criteria for completing reassignment makes more sense than the existing kcontroller criteria, but requiring at least min ISR replicas to be in the ISR (your PR) also seems reasonable.

Reply:
One last question from me -- why do we have math.min(minIsrCount, newTargetReplicas)
Is this covering a case where a non "adding" replica is somehow fallen out of isr but is not under min isr?

(As an aside from your changes, the existing code is a bit strange -- lines 104 and 105 seem to do nothing.)

Reply:
@splett2 Thanks for pointing out the inconsistency. I am not aware of a use case if we want to complete the reassignment before all the replicas in the ISR.
@jolshan The newTargetReplicas is the "adding" replicas and targetIsr can be under min ISR.

Reply:
Hmm

```
            newTargetReplicas = new ArrayList<>(replicas.size());
            for (int replica : replicas) {
                if (!removing.contains(replica)) {
                    newTargetReplicas.add(replica);
                }
            }
```

It seems like it is just the final state -- current replicas minus the removing ones. Not the adding ones?